### ISSUE TITLE
bug: using Terraform functions can and regex to build output hosts list will cause failure

### ISSUE THREAD

**Issue by sean-freeman** [ansible-collections/cloud.terraform#42]
##### SUMMARY
bug: using Terraform functions can and regex to build output hosts list will cause failure

No workaround possible, due to #41 

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME
`cloud.terraform.terraform` Ansible Module

##### ANSIBLE VERSION

```shell
$ ansible --version
ansible [core 2.13.4]
  python version = 3.9.6 (default, Sep  1 2021, 09:17:26) [Clang 12.0.5 (clang-1205.0.22.9)]
  jinja version = 3.1.2
  libyaml = True
```

##### COLLECTION VERSION

```shell
$ ansible-galaxy collection list cloud.terraform
Collection      Version
--------------- -------
cloud.terraform 1.0.0
```

##### STEPS TO REPRODUCE

Adding the following Terraform Output statement to generate the required output format for subsequent execution of Ansible Role `cloud.terraform.inventory_from_outputs`:

```hcl
# https://registry.terraform.io/modules/terraform-aws-modules/ec2-instance/aws/latest
output "host_list" {
  value = [
    for key in module.run_ec2_instance_module: {
      "module_given_host_name" : key.name ,
      "output_host_ip" : key.private_ip ,
      "output_ansible_inventory_group" : can(regex("^db.*",key.module_given_host_name)) ? "databases" : can(regex("^app.*",key.module_given_host_name)) ? can(regex(".*primary.*",key.module_given_host_name)) ? "app_primary" : can(regex(".*secondary.*",key.module_given_host_name)) ? "app_secondary" : "ERROR" : "ERROR" ,
      "output_host_os_user" : "root" ,
    }
  ]
}
```

This was working before:
```yml
  - name: Terraform Template AWS
    register: terraform_template1_result
    environment:
      AWS_ACCESS_KEY_ID: "{{aws_access_key}}"
      AWS_SECRET_ACCESS_KEY: "{{aws_secret_access_key}}"
      AWS_REGION: "{{aws_region}}"
    cloud.terraform.terraform:
      project_path: "{{ playbook_dir }}/terraform_template_aws"
      state: present
      force_init: true
      complex_vars: true
      variables:
        var1: "{{var1}}"
```

##### ACTUAL RESULTS

```shell
TASK [Terraform Template AWS] ****
<127.0.0.1> ESTABLISH LOCAL CONNECTION FOR USER: macos_username
<127.0.0.1> EXEC /bin/sh -c 'echo ~macos_username && sleep 0'
<127.0.0.1> EXEC /bin/sh -c '( umask 77 && mkdir -p "` echo /Users/macos_username/.ansible/tmp `"&& mkdir "` echo /Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373 `" && echo ansible-tmp-1678717392.573625-57544-80416586898373="` echo /Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373 `" ) && sleep 0'

Using module file /Users/macos_username/.ansible/collections/ansible_collections/cloud/terraform/plugins/modules/terraform.py
<127.0.0.1> PUT /Users/macos_username/.ansible/tmp/ansible-local-57489trjv3p9v/tmpirltvusi TO /Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373/AnsiballZ_terraform.py
<127.0.0.1> EXEC /bin/sh -c 'chmod u+x /Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373/ /Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373/AnsiballZ_terraform.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'AWS_ACCESS_KEY_ID=KEY_GOES_HERE AWS_SECRET_ACCESS_KEY=KEY_GOES_HERE AWS_REGION=eu-central-1 /Users/macos_username/.pyenv/versions/3.9.6/bin/python3.9 /Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373/AnsiballZ_terraform.py && sleep 0'
<127.0.0.1> EXEC /bin/sh -c 'rm -f -r /Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373/ > /dev/null 2>&1 && sleep 0'

The full traceback is:
Traceback (most recent call last):
  File "/Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373/AnsiballZ_terraform.py", line 107, in <module>
    _ansiballz_main()
  File "/Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373/AnsiballZ_terraform.py", line 99, in _ansiballz_main
    invoke_module(zipped_mod, temp_path, ANSIBALLZ_PARAMS)
  File "/Users/macos_username/.ansible/tmp/ansible-tmp-1678717392.573625-57544-80416586898373/AnsiballZ_terraform.py", line 47, in invoke_module
    runpy.run_module(mod_name='ansible_collections.cloud.terraform.plugins.modules.terraform', init_globals=dict(_module_fqn='ansible_collections.cloud.terraform.plugins.modules.terraform', _modlib_path=modlib_path),
  File "/Users/macos_username/.pyenv/versions/3.9.6/lib/python3.9/runpy.py", line 210, in run_module
    return _run_module_code(code, init_globals, run_name, mod_spec)
  File "/Users/macos_username/.pyenv/versions/3.9.6/lib/python3.9/runpy.py", line 97, in _run_module_code
    _run_code(code, mod_globals, init_globals,
  File "/Users/macos_username/.pyenv/versions/3.9.6/lib/python3.9/runpy.py", line 87, in _run_code
    exec(code, run_globals)
  File "/var/folders/f5/xv_7b89d7ss448jmc1crwwl40000gn/T/ansible_cloud.terraform.terraform_payload_hb4pncz3/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/modules/terraform.py", line 643, in <module>
  File "/var/folders/f5/xv_7b89d7ss448jmc1crwwl40000gn/T/ansible_cloud.terraform.terraform_payload_hb4pncz3/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/modules/terraform.py", line 575, in main
  File "/var/folders/f5/xv_7b89d7ss448jmc1crwwl40000gn/T/ansible_cloud.terraform.terraform_payload_hb4pncz3/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/terraform_commands.py", line 215, in show
  File "/var/folders/f5/xv_7b89d7ss448jmc1crwwl40000gn/T/ansible_cloud.terraform.terraform_payload_hb4pncz3/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models.py", line 91, in from_json
  File "/var/folders/f5/xv_7b89d7ss448jmc1crwwl40000gn/T/ansible_cloud.terraform.terraform_payload_hb4pncz3/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models.py", line 72, in from_json
  File "/var/folders/f5/xv_7b89d7ss448jmc1crwwl40000gn/T/ansible_cloud.terraform.terraform_payload_hb4pncz3/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models.py", line 73, in <dictcomp>
  File "/var/folders/f5/xv_7b89d7ss448jmc1crwwl40000gn/T/ansible_cloud.terraform.terraform_payload_hb4pncz3/ansible_cloud.terraform.terraform_payload.zip/ansible_collections/cloud/terraform/plugins/module_utils/models.py", line 23, in from_json
KeyError: 'value'
fatal: [localhost]: FAILED! => {
    "changed": false,
    "module_stdout": "",
    "msg": "MODULE FAILURE\nSee stdout/stderr for the exact error",
    "rc": 1
}
```

**Comment by sean-freeman** [comment]
Same error even occurs when using latest code from `main` branch, and using a simplified Terraform Output object.

## Collection upgrade to latest

### Collections file

```yml
collections:

# Collections from public repositories via HTTPS
    version: main
  - name: https://github.com/ansible-collections/cloud.terraform.git
    type: git
    version: main
```

### Collection install from git

```shell
$ ansible-galaxy collection install --requirements-file ansible_requirements_collections.yml

Starting galaxy collection install process
Process install dependency map
Cloning into '/Users/macos_user/.ansible/tmp/ansible-local-61315w3mh3ll8/tmp5dj5y5dv/cloud.terraformfif4x7gv'...
remote: Enumerating objects: 1368, done.
remote: Counting objects: 100% (287/287), done.
remote: Compressing objects: 100% (147/147), done.
remote: Total 1368 (delta 163), reused 197 (delta 124), pack-reused 1081
Receiving objects: 100% (1368/1368), 293.89 KiB | 3.34 MiB/s, done.
Resolving deltas: 100% (451/451), done.
Already on 'main'
Your branch is up to date with 'origin/main'.
Starting collection install process
Installing 'cloud.terraform:1.0.0' to '/Users/macos_user/.ansible/collections/ansible_collections/cloud/terraform'
Created collection for cloud.terraform:1.0.0 at /Users/macos_user/.ansible/collections/ansible_collections/cloud/terraform
cloud.terraform:1.0.0 was installed successfully
```

---

## Simplified Terraform Output code

```hcl
# https://registry.terraform.io/modules/terraform-aws-modules/ec2-instance/aws/latest
output "host_list" {
  value = [
    for key in module.run_ec2_instance_module: {
      "module_given_host_name" : key.name ,
      "output_host_ip" : key.private_ip ,
      "output_ansible_inventory_group" : "all" ,
      "output_host_os_user" : "root" ,
    }
  ]
}
```

**Comment by sean-freeman** [comment]
Same error occurs even when using very basic code

```hcl
# https://registry.terraform.io/modules/terraform-aws-modules/ec2-instance/aws/latest
output "host_list_test" {
  value = [ module.run_ec2_instance_module ]
}
```

**Comment by sstanovnik** [comment]
Thank you for the detailed report. I don't immediately see the cause from the logs, so we'll investigate further and create a test case for a reproducible example.

**Comment by PolonaM** [comment]
@sean-freeman Thank you for the report, but I would need an additional info - can you please add the part from terraform.tfstate where outputs are listed? From the traceback it looks like the format of terraform.tfstate is not as expected

**Comment by sean-freeman** [comment]
@PolonaM The `.tfstate` file is never generated, this fails before the Terraform Plan is generated.

**Comment by PolonaM** [comment]
Can you then manully generate it with "terraform init" and "terraform apply"?

**Comment by sean-freeman** [comment]
@PolonaM  Yes, here is the result of the `.tfstate` file after the Terraform Template was executed with manual `init` then `apply`...

```json
{
  "version": 4,
  "terraform_version": "1.3.7",
  "serial": 226,


  "outputs": {
    "host_list": {
      "value": [
        {
          "output_ansible_inventory_group": "databases",
          "output_host_ip": "10.200.10.215",
          "output_host_name": "db-host",
          "output_host_os_user": "root"
        },
        {
          "output_ansible_inventory_group": "app_primary",
          "output_host_ip": "10.200.10.185",
          "output_host_name": "app-primary",
          "output_host_os_user": "root"
        },
        {
          "output_ansible_inventory_group": "app_secondary",
          "output_host_ip": "10.200.10.84",
          "output_host_name": "app-secondary",
          "output_host_os_user": "root"
        }
      ],
      "type": [
        "tuple",
        [
          [
            "object",
            {
              "output_ansible_inventory_group": "string",
              "output_host_ip": "string",
              "output_host_name": "string",
              "output_host_os_user": "string"
            }
          ],
          [
            "object",
            {
              "output_ansible_inventory_group": "string",
              "output_host_ip": "string",
              "output_host_name": "string",
              "output_host_os_user": "string"
            }
          ],
          [
            "object",
            {
              "output_ansible_inventory_group": "string",
              "output_host_ip": "string",
              "output_host_name": "string",
              "output_host_os_user": "string"
            }
          ]
        ]
      ]
    }
  }


}
```

**Comment by sean-freeman** [comment]
@PolonaM please note, all other issues raised are non-blocking - but this issue is blocking and I hope the fix to the Python code for the above will trigger a `1.0.2` release.

**Comment by PolonaM** [comment]
@sean-freeman Thanks for the sent .tfstate file. The file is as expected, so I can't figure it out, where is the problem. Can you also send me the .tf file so that I can understand what "run_ec2_instance_module" is and that I can reproduce the error?

**Comment by sean-freeman** [comment]
@PolonaM The code block is in the original steps to reproduce, with reference to the Terraform Module > https://registry.terraform.io/modules/terraform-aws-modules/ec2-instance/aws/latest

**Comment by PolonaM** [comment]
@sean-freeman Thanks for the info. From the provided link I was not able to understand what "run_ec2_instance_module" is, but nevertheless I created my own example, where I also used variables in outputs, that will be known after apply (similar to key.private_ip) - and this is what causes the problem, because "terraform plan" will not show the "value" section if inthere are values that will be known after apply. This issue is actually connected to the outputs definition on general, so thanks for discovering the bug, I will open PR with the fix.
